### ISSUE TITLE
the model can not converge using MetricLossOnly example

### ISSUE THREAD

**Issue by j-river** [KevinMusgrave/pytorch-metric-learning#545]
Hi  Kevin,
I tried running the MetricLossOnly notebook, but the model could not converge. I have tried the default ContrastiveLoss, ContrastiveLoss(distance=CosineSimilarity(), pos_margin=1, neg_margin=0), and also tried different pos_margin and neg_margin, not using miner. But it does not work.  Could you please give me some advice?

**Comment by KevinMusgrave** [comment]
Are you running on your own dataset or the example CIFAR dataset?

Have you tried lowering the optimizer's learning rate?

**Comment by j-river** [comment]
I used the cifar 100 dataset and default learning rate,  the lr is 1e-5, seems small enough

**Comment by KevinMusgrave** [comment]
Do you mean the accuracy never gets very high? Like around 10% MAP@R? I don't think I've actually gotten it higher than that in those notebooks.

**Comment by j-river** [comment]
Yes. And I tried the Inference example , try  to find the most similar images  of  the query image by embedding cosin similarity. But      it does not work。 for example ,  the query image is frog, but the searched images are much different.  Besides,  the cosin similarity between imgs is very large , almost 0.98,  no matter these imgs are similar or not.  it seems that the model can not distinguish different images.

**Comment by KevinMusgrave** [comment]
Do you mean you tried the inference example using the model you trained in the `MetricLossOnly` notebook?

**Comment by j-river** [comment]
exactly



| |
17612726179
|
|
***@***.***
|




---- Replied Message ----
| From | Kevin ***@***.***> |
| Date | 11/03/2022 21:38 |
| To | ***@***.***> |
| Cc | ***@***.******@***.***> |
| Subject | Re: [KevinMusgrave/pytorch-metric-learning] the model can not converge using MetricLossOnly example (Issue #545) |

Do you mean you tried the inference example using the model you trained in the MetricLossOnly notebook?

**Comment by KevinMusgrave** [comment]
Some ideas you can try (I might also try later today):

- The model in the `MetricLossOnly` notebook is a ResNet18 pretrained on Imagenet. So it's architecture is for 224x224 images, not 64x64, which is what the images are resized to in the notebook. Using an architecture from here might be better: https://github.com/kuangliu/pytorch-cifar, as long as it's not pretrained on cifar.
- The `RandomResizedCrop` augmentation might be too strong, so removing it might help.
- ClassDisjointCIFAR100 might be difficult. Using the original CIFAR100 will be easier, and might help for debugging.

**Comment by j-river** [comment]
thank you very much!



| |
17612726179
|
|
***@***.***
|




---- Replied Message ----
| From | Kevin ***@***.***> |
| Date | 11/03/2022 22:02 |
| To | ***@***.***> |
| Cc | ***@***.******@***.***> |
| Subject | Re: [KevinMusgrave/pytorch-metric-learning] the model can not converge using MetricLossOnly example (Issue #545) |

Some ideas you can try (I might also try later today):

The model in the MetricLossOnly notebook is a ResNet18 pretrained on Imagenet. So it's architecture is for 224x224 images, not 64x64, which is what the images are resized to in the notebook. Using an architecture from here might be better: https://github.com/kuangliu/pytorch-cifar, as long as it's not pretrained on cifar.
The RandomResizedCrop augmentation might be too strong, so removing it might help.
ClassDisjointCIFAR100 might be difficult. Using the original CIFAR100 will be easier, and might help for debugging.

**Comment by j-river** [comment]
I tried the ideas mentioned above, but got no improvments. I have no idea what is going wrong. Have you tried these ideas and get good results?

**Comment by KevinMusgrave** [comment]
I got better results (28.5 map@r) on the original CIFAR100 dataset, without using `RandomResizedCrop`, and using TripletMarginLoss + MultiSimilarityMiner. I trained for 30 epochs. See the accuracy table and UMAP plot below

|epoch|AMI\_level0|NMI\_level0|mean\_average\_precision\_level0|mean\_average\_precision\_at\_r\_level0|mean\_reciprocal\_rank\_level0|precision\_at\_1\_level0|r\_precision\_level0|best\_epoch|best\_accuracy|
|---|---|---|---|---|---|---|---|---|---|
|5|0\.549604629487344|0\.6039413114650015|0\.16903301084425598|0\.16903301084425598|0\.5961137413978577|0\.4743|0\.29242323232323236|5|0\.16903301084425598|
|10|0\.5917696526194818|0\.641026427369229|0\.21385285745774146|0\.21385285745774146|0\.6276230216026306|0\.5121|0\.33844545454545455|10|0\.21385285745774146|
|15|0\.6126627632865608|0\.6593270282829162|0\.24080940547458968|0\.24080940547458968|0\.6492369771003723|0\.5369|0\.3647373737373738|15|0\.24080940547458968|
|20|0\.6222297443779398|0\.6676764718842069|0\.2568110105737491|0\.2568110105737491|0\.6563907265663147|0\.5485|0\.37933838383838386|20|0\.2568110105737491|
|25|0\.6318022150928577|0\.6760182760960212|0\.27368031266686377|0\.27368031266686377|0\.6668776869773865|0\.5641|0\.3940757575757576|25|0\.27368031266686377|
|30|0\.6363189185876363|0\.6800424647412477|0\.2850454573425632|0\.2850454573425632|0\.6696687340736389|0\.5673|0\.40452828282828285|30|0\.2850454573425632|


Validation set embeddings at epoch 30:
![image](https://user-images.githubusercontent.com/2314891/200382230-f3693d1f-b700-4578-a61b-ea27736630f5.png)

**Comment by j-river** [comment]
I really appreciate your patience and help! Your results are much better. I will try it again.
